syncer/syncer.go

```diff
@@ -1781,7 +1804,9 @@ func (s *Syncer) Run(ctx context.Context) (err error) {
 		// check pass SafeModeExitLoc and try disable safe mode, but not in sharding or replacing error
 		safeModeExitLoc := s.checkpoint.SafeModeExitPoint()
 		if safeModeExitLoc != nil && !s.isReplacingErr && shardingReSync == nil {
-			if binlog.CompareLocation(currentLocation, *safeModeExitLoc, s.cfg.EnableGTID) >= 0 {
+			// TODO: for RowsEvent (in fact other than QueryEvent), `currentLocation` is updated in `handleRowsEvent`
```

Comment:
Will this cause a problem here?

Reply:
originally all `currentLocation` are located in `handleXXX`, to support handle-error, some DDL injections are needed in main routine so for query event `currentLocation` is updated earlier. cc @GMHDBJD 

I'll fix above comments when this CI finishes

Reply:
CI success

Reply:
I have thought this over again. If we update `currentLocation` later after `XIDEvent`, we will open a longer safemode (one transaction). But it won't affect the correctness.